tool/gravity/cli/plan.go

```diff
@@ -87,101 +91,170 @@ func displayOperationPlan(localEnv *localenv.LocalEnvironment, environ LocalEnvi
+	return func() (*storage.OperationPlan, error) {
+		updateEnv, err := environ.NewUpdateEnv()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		plan, err := fsm.GetOperationPlan(updateEnv.Backend, opKey.SiteDomain, opKey.OperationID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		reconciledPlan, err := tryReconcilePlan(context.TODO(), localEnv, updateEnv, *plan)
+		if err != nil {
+			logrus.WithError(err).Warn("Failed to reconcile plan.")
+		} else {
+			plan = reconciledPlan
+		}
+		return plan, nil
 	}
-	err = outputPlan(*plan, format)
-	if err != nil {
-		return trace.Wrap(err)
+}
+
+func displayUpdateOperationPlan(localEnv *localenv.LocalEnvironment, environ LocalEnvironmentFactory, opKey ops.SiteOperationKey, opts displayPlanOptions) error {
+	return outputOrFollowPlan(localEnv, getUpdateOperationPlanFunc(localEnv, environ, opKey), opts)
+}
+
+func getInstallOperationPlanFunc(opKey ops.SiteOperationKey) fsm.GetPlanFunc {
+	return func() (*storage.OperationPlan, error) {
+		plan, err := getPlanFromWizard(opKey)
+		if err == nil {
+			log.Debug("Showing install operation plan retrieved from wizard process.")
+			return plan, nil
+		}
+		return nil, trace.NotFound("could not retrieve plan from wizard process")
 	}
-	return nil
 }
 
-func displayInstallOperationPlan(opKey ops.SiteOperationKey, format constants.Format) error {
-	plan, err := getPlanFromWizard(opKey)
-	if err == nil {
-		log.Debug("Showing install operation plan retrieved from wizard process.")
-		return trace.Wrap(outputPlan(*plan, format))
+func displayInstallOperationPlan(localEnv *localenv.LocalEnvironment, opKey ops.SiteOperationKey, opts displayPlanOptions) error {
+	return outputOrFollowPlan(localEnv, getInstallOperationPlanFunc(opKey), opts)
+}
+
+func getExpandOperationPlanFunc(environ LocalEnvironmentFactory, opKey ops.SiteOperationKey) fsm.GetPlanFunc {
+	return func() (*storage.OperationPlan, error) {
+		joinEnv, err := environ.NewJoinEnv()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer joinEnv.Close()
+		plan, err := fsm.GetOperationPlan(joinEnv.Backend, opKey.SiteDomain, opKey.OperationID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		log.Debug("Showing join operation plan retrieved from local join backend.")
+		return plan, nil
 	}
-	return trace.Wrap(outputPlan(*plan, format))
 }
 
 // displayExpandOperationPlan shows plan of the join operation from the local join backend
-func displayExpandOperationPlan(environ LocalEnvironmentFactory, opKey ops.SiteOperationKey, format constants.Format) error {
-	joinEnv, err := environ.NewJoinEnv()
-	if err != nil {
-		return trace.Wrap(err)
+func displayExpandOperationPlan(localEnv *localenv.LocalEnvironment, environ LocalEnvironmentFactory, opKey ops.SiteOperationKey, opts displayPlanOptions) error {
+	return outputOrFollowPlan(localEnv, getExpandOperationPlanFunc(environ, opKey), opts)
```

Comment:
I guess you could move getXXXPlanFunc inside of the respective displayXXXPlan variant since they're no used elsewhere.

Reply:
That's how I did it originally, but it was a bit messy. I also think having those method in global scope might be a bit more convenient if something/someone ever needs to use them.